src/spec/product-import.spec.coffee

```diff
@@ -1,9 +1,443 @@
+      }
+    ]
+
+sampleReferenceCats =
+  [
+    {
+      id: "category_external_id1"
+    },
+    {
+      id: "category_external_id2"
+    }
+  ]
 
 describe 'ProductImport', ->
 
   beforeEach ->
-    @import = new ProductImport
+    @import = new ProductImport null, Config
 
   it 'should initialize', ->
-    expect(@import).toBeDefined()
\ No newline at end of file
+    expect(@import).toBeDefined()
+    expect(@import.client).toBeDefined()
+    expect(@import.client.constructor.name).toBe 'SphereClient'
+    expect(@import.sync).toBeDefined()
+    expect(@import.sync.constructor.name).toBe 'ProductSync'
+
+
+  describe '::performStream', ->
+
+    it 'should execute callback after finished processing batches', (done) ->
+      spyOn(@import, '_processBatches').andCallFake -> Promise.resolve()
+      @import.performStream [1,2,3], done
+      .catch (err) -> done(_.prettify err)
+
+
+  describe '::_extractUniqueSkus', ->
+
+    it 'should extract 5 unique skus from master and variants', ->
+      skus = @import._extractUniqueSkus(sampleProducts)
+      expect(skus.length).toBe 5
+      expect(skus).toEqual ['a', 'b', 'c', 'd', 'e']
+
+  describe '::_prepareProductFetchBySkuQueryPredicate', ->
+
+    it 'should return predicate with 5 unique skus', ->
+      skus = @import._extractUniqueSkus(sampleProducts)
+      predicate = @import._prepareProductFetchBySkuQueryPredicate(skus)
+      expect(predicate).toEqual 'masterVariant(sku in ("a", "b", "c", "d", "e")) or variants(sku in ("a", "b", "c", "d", "e"))'
+
+  describe '::_isExistingEntry', ->
+
+    it 'should detect existing entries', ->
+      existingProduct = sampleProducts[2]
+      newProduct = sampleProducts[0]
+      expect(@import._isExistingEntry(existingProduct,sampleProductProjectionResponse)).toBeDefined()
+      expect(@import._isExistingEntry(existingProduct,sampleProductProjectionResponse).masterVariant.sku).toEqual "e"
+      expect(@import._isExistingEntry(newProduct,sampleProductProjectionResponse)).toBeUndefined()
+
+  describe '::_resolveReference', ->
+
+    it 'should resolve product type reference and cache the result', (done) ->
+      @import._resetCache()
+      spyOn(@import.client.productTypes, "fetch").andCallFake => Promise.resolve(sampleProductTypeResponse)
+      productTypeRef = { id: 'AGS'}
+      @import._resolveReference(@import.client.productTypes, 'productType', productTypeRef, "name=\"#{productTypeRef.id}\"")
+      .then (result) =>
+        expect(@import.client.productTypes.fetch).toHaveBeenCalled()
+        expect(result).toEqual sampleProductTypeResponse.body.results[0].id
+        expect(@import._cache.productType["AGS"]).toEqual sampleProductTypeResponse.body.results[0].id
+        done()
+      .catch done
+
+    it 'should resolve tax category reference and cache the result', (done) ->
+      @import._resetCache()
+      spyOn(@import.client.taxCategories, "fetch").andCallFake => Promise.resolve(sampleTaxCategoryResponse)
```

Comment:
BTW: you should see all those warnings when you build the source files (or run the lint task)